src/lir.rs

```diff
@@ -0,0 +1,452 @@
+    match exp {
+        Expression::Logical(l) => {
+            counter.ins_number += 1;
+            counter.ins_bytes += mem::size_of::<LogicalExpression>();
+            match l.as_ref() {
+                LogicalExpression::And(l, r) => {
+                    expression_count_heler(l, counter);
+                    expression_count_heler(r, counter);
+                }
+                LogicalExpression::Or(l, r) => {
+                    expression_count_heler(l, counter);
+                    expression_count_heler(r, counter);
+                }
+                LogicalExpression::Not(r) => {
+                    expression_count_heler(r, counter);
+                }
+            }
+        }
+        Expression::Predicate(_p) => {}
+    }
+}
+#[cfg(debug_assertions)]
+impl CountSize for Expression {
+    type Output = ExpressionInsBytes;
+    fn count_size(&self) -> Self::Output {
+        let mut counter = ExpressionInsBytes::new();
+        expression_count_heler(self, &mut counter);
+        counter
+    }
+}
+
+impl Translate for Expression {
+    type Output = LirProgram;
+    fn translate(&self) -> Self::Output {
+        let mut lir = LirProgram::new();
+        lir_translate_helper(self, &mut lir);
+        lir.instructions.shrink_to_fit(); // shrink the memory
+        #[cfg(debug_assertions)]
+        {
+            let ast_counter = self.count_size();
+            ast_counter.ins_bytes;
+            ast_counter.ins_number;
+        }
+        lir
+    }
+}
+
+fn lir_translate_helper(exp: &Expression, lir: &mut LirProgram) {
+    match exp {
+        Expression::Logical(logic_exp) => match logic_exp.as_ref() {
+            LogicalExpression::And(l, r) => {
+                lir.instructions
+                    .push(LirInstruction::LogicalOperator(LirLogicalOperator::And));
+                lir_translate_helper(l, lir);
+                lir_translate_helper(r, lir);
+            }
+            LogicalExpression::Or(l, r) => {
+                lir.instructions
+                    .push(LirInstruction::LogicalOperator(LirLogicalOperator::Or));
+                lir_translate_helper(l, lir);
+                lir_translate_helper(r, lir);
+            }
+            LogicalExpression::Not(r) => {
+                lir.instructions
+                    .push(LirInstruction::LogicalOperator(LirLogicalOperator::Not));
+                lir_translate_helper(r, lir);
+            }
+        },
+        Expression::Predicate(p) => {
+            lir.instructions.push(LirInstruction::Predicate(p.clone()));
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::context::{Context, Match};
+    use crate::interpreter::Execute;
+    use crate::router::Router;
+    use crate::schema::Schema;
+    use crate::semantics::FieldCounter;
+    use std::collections::HashMap;
+    use uuid::Uuid;
+
+    #[derive(Debug, Clone, Copy)]
+    pub enum OperandItem<'a> {
+        Val(bool),
```

Comment:
Never constructed varient? Should us mark it as `dead_code`?

Reply:
fixed.  cleaned up legacy code as it is not needed now.